### PR TITLE
feat(user-notification): Hnipp templates

### PR DESCRIPTION
**Pull request by rafnarnason**
this PR contains 
- added features to support a variety of hnipp templates from contentful
- versioning on endpoints

Attach a link to issue if relevant

## What

Specify what you're trying to achieve

## Why

Specify why you need to achieve this

## Screenshots / Gifs

Attach Screenshots / Gifs to help reviewers understand the scope of the pull request

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [ ] I have rebased against main before asking for a review

### CODE REVIEW

File: apps/services/user-notification/src/app/modules/notifications/notifications.controller.ts [hunk 1]
```diff
@@ -78,4 +87,19 @@ export class NotificationsController {
     this.logger.info('Message queued', { messageId: id, ...message })
     return { id }
   }
+
+  @Get('/test')
+  async test(): Promise<any> {
+    return { date: new Date().toISOString() }
```
Comment (busla) [inline apps/services/user-notification/src/app/modules/notifications/notifications.controller.ts:93]: I think you need to `await` here, might be unrelated to the [pipeline failure issue](https://github.com/island-is/island.is/actions/runs/4065092505/jobs/6999334430), but maybe you can give it a shot?

File: apps/services/user-notification/src/app/modules/notifications/notifications.controller.ts [hunk 2]
```diff
@@ -1,81 +1,138 @@
...
-  throwIfError(await validate(message, { forbidUnknownValues: true }))
-
-  return message
-}
+import { NotificationsService } from './notifications.service'
 
 @Controller('notifications')
-@ApiExtraModels(NewDocumentMessage)
+@ApiExtraModels(CreateNotificationDto)
+// @UseInterceptors(CacheInterceptor) // auto-caching GET responses
```
Comment (kksteini) [inline apps/services/user-notification/src/app/modules/notifications/notifications.controller.ts:33]: Commented out?

File: apps/services/user-notification/src/app/modules/notifications/notifications.service.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+  Injectable,
+} from '@nestjs/common'
+import { CreateHnippNotificationDto } from './dto/createHnippNotification.dto'
+import { HnippTemplate } from './dto/hnippTemplate.response'
+import { Cache } from 'cache-manager'
+
+import axios from 'axios'
+
+// djdjdj dkdkd
+// adding a line due to linting - remove me later
```
Comment (kksteini) [inline apps/services/user-notification/src/app/modules/notifications/notifications.service.ts:16]: Is this still needed.

File: apps/services/user-notification/src/app/modules/notifications/notificationsWorker.service.ts [hunk 4]
```diff
@@ -30,10 +30,10 @@ export class NotificationsWorkerService implements OnApplicationBootstrap {
   }
 
   async run() {
-    await this.worker.run<Message>(
+    await this.worker.run<CreateHnippNotificationDto>(
       async (message, job): Promise<void> => {
         const messageId = job.id
-        this.logger.info('Message received by worker', { messageId })
+        this.logger.info('Message received by worker ... ...', { messageId })
```
Comment (kksteini) [inline apps/services/user-notification/src/app/modules/notifications/notificationsWorker.service.ts:36]: What is this addition for?

File: apps/services/user-notification/src/messages.ts [hunk 5]
```diff
@@ -1,19 +1,19 @@
...
+//         defaultMessage: 'Nýtt skjal á island.is',
+//       },
+//       body: {
+//         id: 'user-notification.messages:new-document-body',
+//         description: 'Body of message letting user know of new inbox document',
+//         defaultMessage: 'Þú átt nýtt skjal á island.is frá: {organization}',
+//       },
+//     }),
+//   },
+// }
```
Comment (kksteini) [inline apps/services/user-notification/src/messages.ts:19]: So, this entire file is commented out?

File: infra/src/uber-charts/islandis.ts [hunk 6]
```diff
@@ -195,8 +195,8 @@ export const FeatureDeploymentServices: ServiceBuilder<any>[] = []
 
 // Services that are included in some environment above but should be excluded from feature deployments
 export const ExcludedFeatureDeploymentServices: ServiceBuilder<any>[] = [
-  userNotificationService,
-  userNotificationWorkerService,
+  // userNotificationService,
+  // userNotificationWorkerService,
```
Comment (kksteini) [inline infra/src/uber-charts/islandis.ts:199]: Keeping commented out code?

File: libs/cms/src/lib/cms.contentful.service.ts [hunk 7]
```diff
@@ -1,3 +1,5 @@
+// TRIGGER LINE FOR FEATURE DEPLOY WORKAROUND - REMOVE BEFORE MERGE
+
```
Comment (kksteini) [inline libs/cms/src/lib/cms.contentful.service.ts:2]: Is this here to make sure the contentful service is deployed and launched? Why isn't that happening automatically?

Comment (rafnarnason) [reply libs/cms/src/lib/cms.contentful.service.ts:2]: no this ... regretfully ... is how triggering a #feature-deploy build works for services such as user-notification ... it´s a hack workaround which infra will address

File: infra/src/uber-charts/islandis.ts (hunk 6 above)
Comment (rafnarnason) [reply infra/src/uber-charts/islandis.ts:199]: commented out during #feature-deploy development - workaround trick from infra

File: apps/services/user-notification/src/messages.ts (hunk 5 above)
Comment (rafnarnason) [reply apps/services/user-notification/src/messages.ts:19]: the whole thing will go away :) new system in place

File: apps/services/user-notification/src/app/modules/notifications/notificationsWorker.service.ts (hunk 4 above)
Comment (rafnarnason) [reply apps/services/user-notification/src/app/modules/notifications/notificationsWorker.service.ts:36]: just a little trick to make sure see which version of worker is receiving the message
the ones with ... ... are the new ones

File: apps/services/user-notification/src/app/modules/notifications/notifications.controller.ts (hunk 2 above)
Comment (rafnarnason) [reply apps/services/user-notification/src/app/modules/notifications/notifications.controller.ts:33]: well spotted - will fix

File: infra/src/uber-charts/islandis.ts (hunk 6 above)
Comment (kksteini) [reply infra/src/uber-charts/islandis.ts]: :+1:

**Review by kksteini — APPROVED** [review]

File: apps/services/user-notification/src/app/modules/notifications/notifications.controller.ts [hunk 8]
```diff
@@ -1,81 +1,138 @@
...
-
-  return message
-}
+import { NotificationsService } from './notifications.service'
 
 @Controller('notifications')
-@ApiExtraModels(NewDocumentMessage)
+@ApiExtraModels(CreateNotificationDto)
+@UseInterceptors(CacheInterceptor) // auto-caching GET responses
 export class NotificationsController {
```
Comment (eirikurn) [inline apps/services/user-notification/src/app/modules/notifications/notifications.controller.ts:34]: Should split this up into multiple versions and REST-ful controllers. E.g.

```
POST /v1/notifications
POST /notifications  - v1 for backwards compatibility.

POST /v2/notifications - new endpoint
GET /v2/templates
GET /v2/templates/:templateId
```

Check out the [docs](https://docs.nestjs.com/techniques/versioning). Can enable versioning in your bootstrap file and then configure version for each controller. Good to also split them up in multiple files and even folders (v1 vs v2).

**Review by eirikurn — APPROVED** [review]
Approving core files (just the build comment).

Skimmed the rest and added some feedback for the API design to better fit our API design guide.

**Pull request merged**